### PR TITLE
RowMasks use bitmasks instead of bitmaps

### PR DESCRIPTION
**Pull request by robert3005**

### CODE REVIEW

File: vortex-file/src/read/mask.rs [hunk 1]
```diff
@@ -1,98 +1,121 @@
...
 }
 
+impl PartialEq for RowMask {
+    fn eq(&self, other: &Self) -> bool {
+        self.begin == other.begin
+            && self.end == other.end
+            && self
+                .bitmask
+                .clone()
+                .into_bool()
```
Comment (lwwmanning) [inline vortex-file/src/read/mask.rs:31]: this is really expensive (going to canonicalize and then throw away) if the array isn't a BoolArray

File: vortex-file/src/read/mask.rs [hunk 2]
```diff
@@ -105,29 +128,35 @@ impl RowMask {
                 return Err(err());
             }
 
-            let mut bitmap = Bitmap::new();
+            let mut buffer_builder = BooleanBufferBuilder::new_from_buffer(MutableBuffer::new((end - begin).div_ceil(8)), end - begin);
```
Comment (lwwmanning) [inline vortex-file/src/read/mask.rs]: `BooleanBufferBuilder::new(end-begin)` will achieve the same effect

File: vortex-file/src/read/mask.rs [hunk 3]
```diff
@@ -105,29 +128,35 @@ impl RowMask {
...
 
             iterate_integer_array!(array, |$P, $iterator| {
                 for batch in $iterator {
                     for index in batch.data() {
-                        bitmap.add(u32::try_from(*index).map_err(|_| err())?);
+                        bit_util::set_bit(buffer_builder.as_slice_mut(), *index as usize)
```
Comment (lwwmanning) [inline vortex-file/src/read/mask.rs]: `buffer_builder.set_bit(*index as usize, true)`

File: vortex-file/src/read/mask.rs (hunk 2 above)
Comment (robert3005) [reply vortex-file/src/read/mask.rs]: Not quite, the code here was wrong

File: vortex-file/src/read/mask.rs (hunk 3 above)
Comment (robert3005) [reply vortex-file/src/read/mask.rs]: the whole idea was that this is always true so I can call underlying method

File: vortex-file/src/read/mask.rs (hunk 1 above)
Comment (robert3005) [reply vortex-file/src/read/mask.rs:31]: yes, this is only for tests. I can make this impl only exist in test mode

**Comment by robert3005** [comment]
~there's still something incorrect with the implementation but I will debug that tomorrow~

File: vortex-file/src/read/mask.rs (hunk 3 above)
Comment (lwwmanning) [reply vortex-file/src/read/mask.rs]: the method is inline anyway

Comment (robert3005) [reply vortex-file/src/read/mask.rs]: fair, I didn't see the annotation

**Comment by danking** [comment]
@robert3005 this looks fine to me. auto-merge is on. is there still something you think is wrong?

**Comment by robert3005** [comment]
Apologies, I have fixed the issues that made CI red

File: vortex-file/src/read/buffered.rs [hunk 4]
```diff
@@ -28,9 +28,7 @@ impl BufferedLayoutReader {
     // TODO(robert): Support out of order reads
     fn buffer_read(&mut self, mask: &RowMask) -> VortexResult<Option<Vec<MessageLocator>>> {
         while let Some(((begin, end), layout)) = self.layouts.pop_front() {
-            if mask.begin() <= begin && begin < mask.end()
-                || mask.begin() < end && end <= mask.end()
-            {
+            if mask.end() > begin && mask.begin() <= end {
```
Comment (lwwmanning) [inline vortex-file/src/read/buffered.rs:31]: this went from "overlapping" to "strictly contains"... why?

Comment (robert3005) [reply vortex-file/src/read/buffered.rs:31]: I think it's the opposite, it went from "strictly contains" to "overlapping". At least that's how I read this logic?

Comment (lwwmanning) [reply vortex-file/src/read/buffered.rs:31]: From closer read... the old one was "the mask's range overlaps at least one of {begin, end}", which would exclude cases where the mask's range was strictly contained by the layout's range (i.e., without overlapping its ends). That does seem obviously wrong in retrospect. The new one is "some amount of the mask range overlaps the layout's (begin, end)" range, which should be correct from how I understand this code

Comment (robert3005) [reply vortex-file/src/read/buffered.rs:31]: I have a thing in #1295 that makes this logic much cleaner but it has other things that I'm not 100% certain of

File: vortex-file/src/read/layouts/test_read.rs [hunk 5]
```diff
@@ -17,9 +16,7 @@ pub fn layout_splits(layout: &mut dyn LayoutReader, length: usize) -> Vec<RowMas
     splits
         .into_iter()
         .tuple_windows::<(usize, usize)>()
-        .map(|(begin, end)| unsafe {
-            RowMask::new_unchecked(Bitmap::from_range(begin as u32..end as u32), 0, end)
-        })
```
Comment (robert3005) [inline vortex-file/src/read/layouts/test_read.rs:22]: this logic was incorrect (didn't match real code) which then masked the bug that I fixed. Need to figure out how to customize regular reader so we don't need this code

**Review by lwwmanning — APPROVED** [review]

**Pull request merged**